File: frontend/src/api/cohort.ts
import { client } from './client'
import type { CohortData, FunnelStep } from '../types'

/** Column mapping for real-data cohort analysis. Omit (or leave partial) to
 * run against the demo snapshot. */
export interface CohortQuery {
  datasource_id?: string | null
  table?: string | null
  entity_col?: string | null
  date_col?: string | null
  stage_col?: string | null
}

export async function retention(cfg: CohortQuery = {}): Promise<CohortData> {
  const { data } = await client.post<CohortData>('/cohort/retention', cfg)
  return data
}

export async function funnel(cfg: CohortQuery = {}): Promise<FunnelStep[]> {
  const { data } = await client.post<{ steps: FunnelStep[] }>('/cohort/funnel', cfg)
  return data.steps
}

File: backend/app/schemas/cohort.py
"""Cohort retention and funnel schemas."""
from __future__ import annotations

from pydantic import BaseModel


class CohortQuery(BaseModel):
    """Column mapping for real-data cohort analysis. All-optional: an empty
    body (or any missing field) falls back to the deterministic demo snapshot."""

    datasource_id: str | None = None
    table: str | None = None
    entity_col: str | None = None
    date_col: str | None = None  # retention: the event timestamp column
    stage_col: str | None = None  # funnel: the stage/event-type column


class CohortCell(BaseModel):
    count: int
    pct: float


class CohortResponse(BaseModel):
    cohorts: list[str]  # cohort month labels, e.g. "2024-03"
    offsets: list[int]  # month offsets 0..N
    sizes: list[int]  # cohort sizes, parallel to `cohorts`
    cells: list[list[CohortCell | None]]  # None = beyond the data range


class FunnelStep(BaseModel):
    name: str
    count: int
    pct_of_first: float
    drop_pct: float


class FunnelResponse(BaseModel):
    steps: list[FunnelStep]
